### DISCUSSION TITLE
fail2ban use case with a lot of failures and few bans (-> purging of failList)

### DISCUSSION THREAD

**Discussion by krskrs** [fail2ban/fail2ban#2945] [General]
Hi @sebres , first off, let me thank you for the very good job you're doing with fail2ban, and for the kind support you're providing. :)

This said, I was considering to use fail2ban in a perhaps unusual way, to blindly count all requests to a web site backend, to ban the users making too many requests in a short time span (ie: malicious crawlers which copy contents, etc.).

So, such use case would essentially translate to: a moderately short findTime and a big number of failures being collected, and a **small number of bans**.

Being myself a developer, I tried to pinpoint the weak points of this approach, and I explored fail2ban code to check how failures were being cleaned up after going outside of the findTime time window. I can be wrong, as today is the first time I have looked at fail2ban code, but so far I have found out that the __failList in FailManager is only purged when a ban is triggered? Is this correct? If so, I'm afraid that such list could keep growing too much in my use case, and perhaps a patch would be needed.

What do you think? Can you confirm and/or do you have any suggestions/recommendations? Before deciding to try fail2ban and exploring the code, I had been searching for information on the internet and I had found some SO threads where some people suggested that fail2ban could be used this way, so maybe this could be something of general interests.

**Comment by sebres** [comment]
> I was considering to use fail2ban in a perhaps unusual way, to blindly count all requests to a web site backend, to ban the users making too many requests in a short time span (ie: malicious crawlers which copy contents, etc.)

Well, I would not say it is unusual way directly, but it may be indeed too heavy. Take a look in our wiki for [Best practice](https://github.com/fail2ban/fail2ban/wiki/Best-practice) as well as for filter like [nginx-limit-req](https://github.com/fail2ban/fail2ban/blob/master/config/filter.d/nginx-limit-req.conf) which would allow better suitable solution for limiting requests and banning such crawlers.

> I have found out that the __failList in FailManager is only purged when a ban is triggered?

Hmm... This may be indeed an issue (as for cleanup which is indeed the major function that lets shrink a failure list, there are other routines but affecting only certain tickets).
I think e3737bb7c095f7c029e1b087027995653db4535c in attempt to fix a race condition partially introduced another one, which is affected only if banASAP is True, but unfortunately it is always the case in stock fail2ban version.
I'll check what we can do here and provide a fix for the issue, Thx!

**Reply by sebres** [reply]
> do you have any suggestions/recommendations?

Also note options from RFE #2402 which could help to reduce memory usage on jails with a lot of long-life failures.

**Reply by krskrs** [reply]
Hi @sebres ! Thank you for following up and for your prompt interest in the matter!

I have already started to read and check all notes and code from the links you shared.  I still have to carefully go through the code related to banASAP, although I remember that yesterday I had already found such parts, but just had a quick glance, and didn't notice anything relevant regarding how failures are pruned without a ban.

This said, pruning more aggressively, or not even storing log lines through the maxmatches options you mentioned, will indeed help to save memory when I will setup fail2ban on the live environment, but my biggest worry now is that, as I foresee long (very long) intervals without a single ban, it will keep storing every single ip address it encounters and memory will become  a problem. Moreover it will have to prune a huge list all together, on first ban, but in the end this is really a secondary concern. The real one is the former.

So, I like very much this program, and from what I have seen in the code, it is already well optimized and I'm quite convinced :) that its code base is already  flexible enough to allow to handle my use case without needing too many changes. Bottom line, I believe that finding a good place for pruning 'expired' failures at regular intervals, or, if not strictly regular, not even only depending on bans, will already be a very good start. 

I'm available, and will be glad to, if needed, to do some beta testing on live environment of any patches we might want to try, and to closely monitor resource usages, etc. My aim is to run it on a VPS with 1GB ram with around 15/20k daily unique ips per day, which will be practically all matched by the failRegex, with the risk of also having days without a single ban.

**Reply by krskrs** [reply]
I believe I found what you were referring to.. 😃  So, in the main loop of the FilterPoll thread, before introducing banASAP, there was this part where performBan() was being called without the ip, which led the subsequent call to failmanager.toBan to iterate all failList looking for possible ban candidates, and eventually it raised the exception which led to the cleanup code!

Well, this part, which so far I had missed, seems to be exactly  what I had in mind and was talking of in my last comment. And I'm happy to know that such code has been in production, and I suppose, wasn't creating too much harm. Practically, cleanup code was being executed every time a modification was detected in the log file? 

Now we just have to find a compromise to make both optimizations coexist, or perhaps even introduce some dedicate configuration option in case some tweaking could be useful in different situations / use cases.


```Python
if self.__modified:
	if not self.banASAP: # pragma: no cover
		self.performBan()
	self.__modified = False
```

```Python
def performBan(self, ip=None):
"""Performs a ban for IPs (or given ip) that are reached maxretry of the jail."""
try: # pragma: no branch - exception is the only way out
	while True:
		ticket = self.failManager.toBan(ip)
		self.jail.putFailTicket(ticket)
except FailManagerEmpty:
	self.failManager.cleanup(MyTime.time())
```

```Python
def toBan(self, fid=None):
with self.__lock:
	for fid in ([fid] if fid is not None and fid in self.__failList else self.__failList):
		data = self.__failList[fid]
		if data.getRetry() >= self.__maxRetry:
			del self.__failList[fid]
			return data
self.__bgSvc.service()
raise FailManagerEmpty
```

**Comment by sebres** [comment]
There is new branch [fix-rc-on-too-many-failures](https://github.com/fail2ban/fail2ban/compare/0.10...sebres:fix-rc-on-too-many-failures) now (rebased from my experimental branch, where this banASAP stuff is completely removed) and it has no RC for cleanup process. 
I needed to adjust several test-cases, because ban ASAP rule will be also followed in test suite now, so failure reached `maxretry` causes immediate ban now.
The cleanup is moved to new method `performSvc`, which is called every 10 tick (up to 20 seconds in quiet jails) or after each ban (as previously) and after each 100 processed lines, but it will invoke cleanup of failure list not earlier than 5 seconds after last check. 
So basically it does not matter how often `performSvc` gets called. Therefore this can be also considered as a certain performance enhancement (will avoid clean up too often).
I have another variant of failmanager implementing failList as a trie with grouping by subnet etc (which is basically part of another RFE), but at once it allows direct selection by time, because the tickets are ordered in this queue, so if this gets ready, this 5 second latency gets replaced (due to last known time of first failure to remove).

**Reply by krskrs** [reply]
Great! Later I will review all of your changes, I'm *really* curious and write some notes here! Thanks!!

**Reply by krskrs** [reply]
So, I went through all changes, and, well, you have indeed taken the best from the 2 concepts and put them together (the banning ASAP concept, and an efficient and good pruning of expired tickets).
All worries I had about the particular use case I'm willing to test are now gone with current implementation, and I'm looking forward to testing in production.
Now I've seen that, as you also said, you've based these changes on an experimental branch which is missing some commits compared to the latest stock 0.11.2 version
How would you suggest me to proceed? I mean, do you think I should wait for you to merge these new changes in a more up to date branch? Or do you believe I should try to directly install this experimental branch because the changes from the newest commits, which are missing here, are not critical?
Thank you again...

**Reply by sebres** [reply]
No, I rebased it from experimental branch to 0.10 branch (where the fix basically belongs), which of course don't have all the features of 0.11th branch. But I'll merge it in 0.11 and 1.0 too.

**Reply by sebres** [reply]
Here you go - fb6315ea5eb4db7e3a7fe460a2c3316926c177af is your merge point of 0.11 branch

**Reply by krskrs** [reply]
Thank you! Very appreciated..

Tonight I'm going to start testing! Will keep you updated.
